src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -187,123 +187,130 @@ state switch
+                // If we're already finalized then just return what we have, and with the frozen bit flipped so that
+                // any future forks keep things frozen.
+                if (state is FinalCompilationTrackerState finalState)
+                {
+                    var frozenFinalState = finalState.WithIsFrozen();
+                    return new CompilationTracker(this.ProjectState, frozenFinalState, this.SkeletonReferenceCache.Clone());
+                }
+                else
+                {
+                    // Transition us to a frozen in progress state.  With the best compilations up to this point, but no
+                    // more pending actions, and with the frozen bit flipped so that any future forks keep things
+                    // frozen.
+                    var inProgressState = InProgressState.Create(
+                        isFrozen: true,
+                        compilationPair.CompilationWithoutGeneratedDocuments,
+                        generatorInfo,
+                        compilationPair.CompilationWithGeneratedDocuments,
+                        ImmutableList<(ProjectState, CompilationAndGeneratorDriverTranslationAction)>.Empty);
+
+                    return new CompilationTracker(inProgressProject, inProgressState, this.SkeletonReferenceCache.Clone());
+                }
             }
 
-            private ICompilationTracker FreezePartialStateWorker(
-                SolutionCompilationState compilationState,
-                DocumentState? docState,
-                SyntaxTree? tree,
+            public ICompilationTracker FreezePartialStateWithDocument(
+                DocumentState docState,
                 CancellationToken cancellationToken)
             {
                 GetPartialCompilationState(
-                    compilationState,
-                    docState?.Id,
+                    docState.Id,
+                    out var state,
                     out var inProgressProject,
                     out var compilationPair,
                     out var generatorInfo,
-                    out var metadataReferenceToProjectId,
                     cancellationToken);
 
-                if (docState == null)
+                // If we're already finalized, and no change has been made to this document.  Then just return what we
+                // have (just with the frozen bit flipped so that any future forks keep things frozen).
+                if (state is FinalCompilationTrackerState finalState &&
+                    this.ProjectState.DocumentStates.TryGetState(docState.Id, out var oldState) &&
+                    oldState == docState)
                 {
-                    Contract.ThrowIfTrue(tree != null);
+                    var frozenFinalState = finalState.WithIsFrozen();
+                    return new CompilationTracker(this.ProjectState, frozenFinalState, this.SkeletonReferenceCache.Clone());
                 }
                 else
```

Comment:
>else [](http://example.com/codeflow?start=16&length=4)

Out of curiosity, why did you add this in the last commit?

Reply:
Just a style preference. I preferred the indentation